# Для тренировки модели нужно установить turicreate
Linux or mac
python 3.6 - 3.8

`pip install turicreate`

### На вход:
- _attend_features.csv (с нашими фичами)
- _groups.csv

### Выход:
- predicts/predict_v01.csv


In [1]:
import turicreate as tc
import pandas as pd

In [2]:
actions = tc.SFrame.read_csv('_attend_features.csv')

Finished parsing file /media/vasiliev/SSD_Data/_dev/ezee_may_2003/ds/_attend_features.csv

Parsing completed. Parsed 100 lines in 0.215888 secs.

------------------------------------------------------
Inferred types from first 100 line(s) of file as 
column_type_hints=[int,int,int,float,str,str,str,float,str,str,str,str,float,float,str,int,str,float,float]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------


Read 173459 lines. Lines per second: 617162

Finished parsing file /media/vasiliev/SSD_Data/_dev/ezee_may_2003/ds/_attend_features.csv

Parsing completed. Parsed 600390 lines in 0.905295 secs.

### Делим общую таблицу на фичи для attend, users, groups

In [3]:
print(actions.head)
actions1 = actions[['group_id',
                    'user_id',
                    'log_visits',
                    #'user_lat',
                    #'user_long',
                    'направление 1',
                    'направление 2',
                    #'направление 3',
                    'weekend']]

users_1 = actions[['user_id',
                   'sex','age',
                   'user_address',
                   'user_lat',
                   'user_long']]

groups_1 = actions[['group_id',
                    'lat',
                    'long',
                    'направление 1',
                    'направление 2',
                    'направление 3',
                    'is_online',
                    'weekend',
                    'lesson_length',
                    'session']]

<bound method SFrame.head of Columns:
	X1	int
	group_id	int
	user_id	int
	log_visits	float
	направление 3	str
	is_online	str
	weekend	str
	lesson_length	float
	session	str
	направление 1	str
	направление 2	str
	group_address	str
	lat	float
	long	float
	sex	str
	age	int
	user_address	str
	user_lat	float
	user_long	float

Rows: 600390

Data:
+----+-----------+-----------+--------------------+---------------+-----------+
| X1 |  group_id |  user_id  |     log_visits     | направление 3 | is_online |
+----+-----------+-----------+--------------------+---------------+-----------+
| 0  | 801346549 | 101370217 | 2.0794415416798357 |  Здорово жить |  offline  |
| 1  | 801346549 | 101375025 | 3.044522437723423  |  Здорово жить |  offline  |
| 2  | 801346549 | 101392971 |        0.0         |  Здорово жить |  offline  |
| 3  | 801346550 | 101352023 | 4.1588830833596715 |  ОНЛАЙН Цигун |   online  |
| 4  | 801346550 | 101354499 | 4.1588830833596715 |  ОНЛАЙН Цигун |   online  |
| 5  | 801346550 |

There are a variety of machine learning techniques that can be used to build a recommender model. Turi Create provides a method turicreate.recommender.create that will automatically choose an appropriate model for your data set. More details on choosing a model: https://apple.github.io/turicreate/docs/userguide/recommender/choosing-a-model.html

First we create a random split of the data to produce a validation set that can be used to evaluate the model.

### Чему учим модель

Мы учим модель предсказывать логарифм от количества посещений юзером курса (логарифм потому что разброс посещений курса пользователем высокий - от 1 до 64)

### Создание и обучение дефолтной модели

In [4]:
training_data, validation_data = tc.recommender.util.random_split_by_user(actions1, 'user_id', 'group_id', item_test_proportion = 0.01)

model = tc.recommender.create(training_data,
                              'user_id',
                              'group_id',
                              target = 'log_visits',
                              user_data = users_1,
                              item_data = groups_1
                             )


Preparing data set.

Data has 600249 observations with 50906 users and 24535 items.

Data prepared in: 0.719134s

Training ranking_factorization_recommender for recommendations.

+--------------------------------+--------------------------------------------------+----------+

| Parameter                      | Description                                      | Value    |

+--------------------------------+--------------------------------------------------+----------+

| num_factors                    | Factor Dimension                                 | 32       |

| regularization                 | L2 Regularization on Factors                     | 1e-09    |

| solver                         | Solver used for training                         | adagrad  |

| linear_regularization          | L2 Regularization on Linear Coefficients         | 1e-09    |

| ranking_regularization         | Rank-based Regularization Weight                 | 0.25     |

| side_data_factorization        | Assign Factors for Side Data                     | True     |

| max_iterations                 | Maximum Number of Iterations                     | 25       |

+--------------------------------+--------------------------------------------------+----------+

Optimizing model using SGD; tuning step size.

Using 75031 / 600249 points for tuning the step size.

+---------+-------------------+------------------------------------------+

| Attempt | Initial Step Size | Estimated Objective Value                |

+---------+-------------------+------------------------------------------+

| 0       | 2.63158           | Not Viable                               |

| 1       | 0.657895          | Not Viable                               |

| 2       | 0.164474          | Not Viable                               |

| 3       | 0.0411184         | 0.30167                                  |

| 4       | 0.0205592         | 0.401023                                 |

| 5       | 0.0102796         | 0.713983                                 |

+---------+-------------------+------------------------------------------+

| Final   | 0.0411184         | 0.30167                                  |

+---------+-------------------+------------------------------------------+

Starting Optimization.

+---------+--------------+-------------------+-----------------------+-------------+

| Iter.   | Elapsed Time | Approx. Objective | Approx. Training RMSE | Step Size   |

+---------+--------------+-------------------+-----------------------+-------------+

| Initial | 63us         | 2.77714           | 1.19031               |             |

+---------+--------------+-------------------+-----------------------+-------------+

| 1       | 674.018ms    | 1.36666           | 1.05557               | 0.0411184   |

| 2       | 1.24s        | 0.865109          | 0.830994              | 0.0411184   |

| 3       | 1.77s        | 0.756861          | 0.769095              | 0.0411184   |

| 4       | 2.27s        | 0.704674          | 0.73634               | 0.0411184   |

| 5       | 2.79s        | 0.672893          | 0.715647              | 0.0411184   |

| 10      | 5.33s        | 0.596647          | 0.662385              | 0.0411184   |

| 20      | 10.36s       | 0.530557          | 0.614437              | 0.0411184   |

| 25      | 12.87s       | 0.508622          | 0.597776              | 0.0411184   |

+---------+--------------+-------------------+-----------------------+-------------+

Optimization Complete: Maximum number of passes through the data reached.

Computing final objective value and training RMSE.

Final objective value: 0.48189

Final training RMSE: 0.577227

In [5]:
# параметры модели
model

Class                            : RankingFactorizationRecommender

Schema
------
User ID                          : user_id
Item ID                          : group_id
Target                           : log_visits
Additional observation features  : 3
User side features               : ['user_id', 'sex', 'age', 'user_address', 'user_lat', 'user_long']
Item side features               : ['group_id', 'lat', 'long', 'направление 1', 'направление 2', 'направление 3', 'is_online', 'weekend', 'lesson_length', 'session']

Statistics
----------
Number of observations           : 600249
Number of users                  : 50906
Number of items                  : 24535

Training summary
----------------
Training time                    : 16.3295

Model Parameters
----------------
Model class                      : RankingFactorizationRecommender
num_factors                      : 32
binary_target                    : 0
side_data_factorization          : 1
solver                           : auto
n

### Получение предсказаний от дефолтной модели

In [6]:
default_recommendations = model.recommend()

recommendations finished on 1000/50906 queries. users per second: 846.034

recommendations finished on 2000/50906 queries. users per second: 848.58

recommendations finished on 3000/50906 queries. users per second: 851.569

recommendations finished on 4000/50906 queries. users per second: 852.843

recommendations finished on 5000/50906 queries. users per second: 853.684

recommendations finished on 6000/50906 queries. users per second: 854.359

recommendations finished on 7000/50906 queries. users per second: 854.626

recommendations finished on 8000/50906 queries. users per second: 854.492

recommendations finished on 9000/50906 queries. users per second: 854.897

recommendations finished on 10000/50906 queries. users per second: 854.943

recommendations finished on 11000/50906 queries. users per second: 855.071

recommendations finished on 12000/50906 queries. users per second: 854.997

recommendations finished on 13000/50906 queries. users per second: 854.39

recommendations finished on 14000/50906 queries. users per second: 854.811

recommendations finished on 15000/50906 queries. users per second: 854.615

recommendations finished on 16000/50906 queries. users per second: 854.443

recommendations finished on 17000/50906 queries. users per second: 854.462

recommendations finished on 18000/50906 queries. users per second: 854.537

recommendations finished on 19000/50906 queries. users per second: 854.585

recommendations finished on 20000/50906 queries. users per second: 854.691

recommendations finished on 21000/50906 queries. users per second: 854.93

recommendations finished on 22000/50906 queries. users per second: 855.113

recommendations finished on 23000/50906 queries. users per second: 855.314

recommendations finished on 24000/50906 queries. users per second: 855.454

recommendations finished on 25000/50906 queries. users per second: 855.534

recommendations finished on 26000/50906 queries. users per second: 855.649

recommendations finished on 27000/50906 queries. users per second: 855.876

recommendations finished on 28000/50906 queries. users per second: 856.048

recommendations finished on 29000/50906 queries. users per second: 856.359

recommendations finished on 30000/50906 queries. users per second: 856.476

recommendations finished on 31000/50906 queries. users per second: 856.439

recommendations finished on 32000/50906 queries. users per second: 856.563

recommendations finished on 33000/50906 queries. users per second: 856.601

recommendations finished on 34000/50906 queries. users per second: 856.663

recommendations finished on 35000/50906 queries. users per second: 856.68

recommendations finished on 36000/50906 queries. users per second: 856.855

recommendations finished on 37000/50906 queries. users per second: 856.953

recommendations finished on 38000/50906 queries. users per second: 857.087

recommendations finished on 39000/50906 queries. users per second: 857.046

recommendations finished on 40000/50906 queries. users per second: 857.058

recommendations finished on 41000/50906 queries. users per second: 857.105

recommendations finished on 42000/50906 queries. users per second: 857.072

recommendations finished on 43000/50906 queries. users per second: 857.144

recommendations finished on 44000/50906 queries. users per second: 857.185

recommendations finished on 45000/50906 queries. users per second: 856.118

recommendations finished on 46000/50906 queries. users per second: 853.077

recommendations finished on 47000/50906 queries. users per second: 834.637

recommendations finished on 48000/50906 queries. users per second: 813.178

recommendations finished on 49000/50906 queries. users per second: 793.124

recommendations finished on 50000/50906 queries. users per second: 774.359

### Посмотрим рекоммендации для отдельных юзеров

In [7]:
df_groups = pd.read_csv('groups.csv')

def display_history_and_predictions_for_user(user_id, train_dataset, predictions):
    print('----------------------------------------------------------------------------')
    print('User', user_id, 'history:')
    display(train_dataset[train_dataset['user_id'] == user_id][
        'group_id','is_online','session','направление 1','направление 2'])
    predictions = predictions[predictions['user_id'] == user_id].to_dataframe()
    print('User recommendations:')
    predicted_groups = predictions['group_id'].to_list()
    display(df_groups[df_groups['уникальный номер'].isin(predicted_groups)].drop(columns=[
        'адрес площадки','расписание в активных периодах','расписание в закрытых периодах','расписание в плановом периоде']))

In [8]:
user_id = 101352882

display_history_and_predictions_for_user(user_id, actions, default_recommendations)

----------------------------------------------------------------------------
User 101352882 history:


Columns:
	group_id	int
	is_online	str
	session	str
	направление 1	str
	направление 2	str

Rows: Unknown

Data:
+-----------+-----------+---------+-----------------------+
|  group_id | is_online | session |     направление 1     |
+-----------+-----------+---------+-----------------------+
| 801348234 |   online  |  mornig | Физическая активность |
| 801348838 |  offline  |  mornig | Физическая активность |
| 801348944 |  offline  |  mornig |         Танцы         |
| 801349048 |  offline  |  mornig | Физическая активность |
| 801350227 |   online  |  mornig |       Творчество      |
| 801350380 |   online  |  mornig |         Танцы         |
| 801351413 |   online  |  mornig |       Рисование       |
| 801358464 |   online  |  mornig | Физическая активность |
| 801358465 |   online  |  mornig |         Танцы         |
| 801358466 |   online  |  mornig | Физическая активность |
+-----------+-----------+---------+-----------------------+
+-------------------------------+
|         направление 2         |
+-------------------------------+
|       ОНЛАЙН Гимнастика       |
|           Гимнастика          |
|             Танцы             |
|              ОФП              |
| ОНЛАЙН Художественно-прикл... |
|          ОНЛАЙН Танцы         |
|        ОНЛАЙН Рисование       |
|       ОНЛАЙН Гимнастика       |
|          ОНЛАЙН Танцы         |
|       ОНЛАЙН Гимнастика       |
+-------------------------------+
[? rows x 5 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

User recommendations:


уникальный номер          направление 1      направление 2   
1008         801354484              Рисование   ОНЛАЙН Рисование  \
1093         801346701              Рисование   ОНЛАЙН Рисование   
1840         801352851              Рисование   ОНЛАЙН Рисование   
1841         801348029              Рисование   ОНЛАЙН Рисование   
2920         801350775              Рисование   ОНЛАЙН Рисование   
2924         801352907              Рисование   ОНЛАЙН Рисование   
2927         801352685              Рисование   ОНЛАЙН Рисование   
3752         801347074  Физическая активность  ОНЛАЙН Гимнастика   
3915         801348042              Рисование   ОНЛАЙН Рисование   
3971         801350767              Рисование   ОНЛАЙН Рисование   

                                 направление 3   
1008                                ОНЛАЙН ИЗО  \
1093                  ОНЛАЙН Масляная живопись   
1840           ОНЛАЙН Китайская живопись У-Син   
1841           ОНЛАЙН Китайская живопись У-Син   
2920           ОНЛАЙН Китайская живопись У-Син   
2924           ОНЛАЙН Китайская живопись У-Син   
2927        ОНЛАЙН Различные техники рисования   
3752  ОНЛАЙН Корригирующая гимнастика для глаз   
3915                  ОНЛАЙН Масляная живопись   
3971        ОНЛАЙН Различные техники рисования   

                             округ площадки   
1008   Юго-Восточный административный округ  \
1093       Восточный административный округ   
1840       Восточный административный округ   
1841   Юго-Восточный административный округ   
2920                                    NaN   
2924                                    NaN   
2927                                    NaN   
3752                                    NaN   
3915  Зеленоградский административный округ   
3971                                    NaN   

                              район площадки  
1008             муниципальный округ Люблино  
1093  муниципальный округ Северное Измайлово  
1840    муниципальный округ Косино-Ухтомский  
1841          муниципальный округ Некрасовка  
2920                                     NaN  
2924                                     NaN  
2927                                     NaN  
3752                                     NaN  
3915             муниципальный округ Крюково  
3971                                     NaN

In [9]:
user_id = 101354253

display_history_and_predictions_for_user(user_id, actions, default_recommendations)

----------------------------------------------------------------------------
User 101354253 history:


Columns:
	group_id	int
	is_online	str
	session	str
	направление 1	str
	направление 2	str

Rows: Unknown

Data:
+-----------+-----------+-----------+-----------------------+-------------------+
|  group_id | is_online |  session  |     направление 1     |   направление 2   |
+-----------+-----------+-----------+-----------------------+-------------------+
| 801347802 |   online  |   mornig  | Физическая активность |     ОНЛАЙН ОФП    |
| 801348667 |   online  |  evening  | Физическая активность | ОНЛАЙН Гимнастика |
| 801348794 |   online  |   mornig  | Физическая активность | ОНЛАЙН Гимнастика |
| 801349796 |  offline  |   mornig  | Физическая активность |     Гимнастика    |
| 801349875 |  offline  | aftennoon | Физическая активность |     Гимнастика    |
| 801350556 |   online  |   mornig  | Физическая активность |     ОНЛАЙН ОФП    |
| 801350568 |   online  |   mornig  | Физическая активность |     ОНЛАЙН ОФП    |
| 801361065 |  offline  |   mornig  |       Рисование       |     Рисование     |
+-----------+-----------+-----------+-----------------------+-------------------+
[? rows x 5 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

User recommendations:


уникальный номер                  направление 1        направление 2   
1076          801349915          Физическая активность           ОНЛАЙН ОФП  \
1613          801348774          Физическая активность    ОНЛАЙН Гимнастика   
1784          801348369          Физическая активность    ОНЛАЙН Гимнастика   
2897          801346717          Физическая активность    ОНЛАЙН Гимнастика   
2899          801346566          Физическая активность    ОНЛАЙН Гимнастика   
2909          801347778          Физическая активность    ОНЛАЙН Гимнастика   
2911          801348554          Физическая активность    ОНЛАЙН Гимнастика   
3730          801346573          Физическая активность           ОНЛАЙН ОФП   
11484         801352857          Физическая активность           Гимнастика   
23659         801373704  Центры московского долголетия  Свободное посещение   

                                         направление 3   
1076                                        ОНЛАЙН ОФП  \
1613                                      ОНЛАЙН Цигун   
1784                                 ОНЛАЙН Гимнастика   
2897                     ОНЛАЙН Дыхательная гимнастика   
2899                     ОНЛАЙН Дыхательная гимнастика   
2909                     ОНЛАЙН Дыхательная гимнастика   
2911                     ОНЛАЙН Дыхательная гимнастика   
3730                                        ОНЛАЙН ОФП   
11484  Дыхательная гимнастика по системе Стрельниковой   
23659                              Свободное посещение   

                                          округ площадки   
1076   Северо-Восточный административный округ, Север...  \
1613   Юго-Западный административный округ, Юго-Восто...   
1784                    Восточный административный округ   
2897                                                 NaN   
2899                                                 NaN   
2909                                                 NaN   
2911                                                 NaN   
3730   Северный административный округ, Северный адми...   
11484            Северо-Восточный административный округ   
23659                   Восточный административный округ   

                                          район площадки  
1076   муниципальный округ Марьина Роща, муниципальны...  
1613   муниципальный округ Южное Бутово, муниципальны...  
1784              муниципальный округ Северное Измайлово  
2897                                                 NaN  
2899                                                 NaN  
2909                                                 NaN  
2911                                                 NaN  
3730   муниципальный округ Дмитровский, муниципальный...  
11484                      муниципальный округ Ростокино  
23659                     муниципальный округ Сокольники

In [10]:
user_id = 101385865

display_history_and_predictions_for_user(user_id, actions, default_recommendations)

----------------------------------------------------------------------------
User 101385865 history:


Columns:
	group_id	int
	is_online	str
	session	str
	направление 1	str
	направление 2	str

Rows: Unknown

Data:
+-----------+-----------+-----------+-----------------------+
|  group_id | is_online |  session  |     направление 1     |
+-----------+-----------+-----------+-----------------------+
| 801346923 |   online  |   mornig  |         Танцы         |
| 801348309 |   online  |   mornig  |       Творчество      |
| 801348946 |   online  |   mornig  | Физическая активность |
| 801349093 |   online  |   mornig  |      Образование      |
| 801350488 |  offline  | aftennoon | Физическая активность |
| 801350739 |  offline  |   mornig  |         Танцы         |
| 801350759 |  offline  |   mornig  |         Танцы         |
| 801351106 |   online  |   mornig  |       Творчество      |
| 801351519 |  offline  |   mornig  |         Танцы         |
| 801351888 |  offline  |   mornig  |         Танцы         |
+-----------+-----------+-----------+-----------------------+
+--------------------------+
|      направление 2       |
+--------------------------+
|       ОНЛАЙН Танцы       |
|  ОНЛАЙН Красота и стиль  |
|    ОНЛАЙН Гимнастика     |
| ОНЛАЙН Иностранные языки |
|    Фитнес, тренажеры     |
|          Танцы           |
|          Танцы           |
|  ОНЛАЙН Красота и стиль  |
|          Танцы           |
|          Танцы           |
+--------------------------+
[? rows x 5 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

User recommendations:


уникальный номер          направление 1   
1307          801347350  Физическая активность  \
1744          801350068             Творчество   
2144          801346784            Образование   
2527          801352390  Физическая активность   
3557          801346557            Образование   
3752          801347074  Физическая активность   
3788          801347329  Физическая активность   
9810          801351000  Физическая активность   
9812          801348614  Физическая активность   
11156         801349550  Физическая активность   

                                направление 2   
1307                        ОНЛАЙН Гимнастика  \
1744                   ОНЛАЙН Красота и стиль   
2144   ОНЛАЙН История, искусство, краеведение   
2527                        ОНЛАЙН Гимнастика   
3557         ОНЛАЙН Психология и коммуникации   
3752                        ОНЛАЙН Гимнастика   
3788                        ОНЛАЙН Гимнастика   
9810                        ОНЛАЙН Гимнастика   
9812                        ОНЛАЙН Гимнастика   
11156                       ОНЛАЙН Гимнастика   

                                           направление 3   
1307            ОНЛАЙН Корригирующая гимнастика для глаз  \
1744   ОНЛАЙН Мастер-класс по уходу за кожей в зрелом...   
2144                     ОНЛАЙН История, культура России   
2527            ОНЛАЙН Корригирующая гимнастика для глаз   
3557                     ОНЛАЙН Психологические тренинги   
3752            ОНЛАЙН Корригирующая гимнастика для глаз   
3788            ОНЛАЙН Корригирующая гимнастика для глаз   
9810            ОНЛАЙН Корригирующая гимнастика для глаз   
9812            ОНЛАЙН Корригирующая гимнастика для глаз   
11156           ОНЛАЙН Корригирующая гимнастика для глаз   

                              округ площадки   
1307        Восточный административный округ  \
1744   Юго-Западный административный округ,    
2144                                     NaN   
2527                                     NaN   
3557                                     NaN   
3752                                     NaN   
3788        Восточный административный округ   
9810                                     NaN   
9812                                     NaN   
11156                                    NaN   

                               район площадки  
1307   муниципальный округ Северное Измайлово  
1744            муниципальный округ Ясенево,   
2144                                      NaN  
2527                                      NaN  
3557                                      NaN  
3752                                      NaN  
3788   муниципальный округ Северное Измайлово  
9810                                      NaN  
9812                                      NaN  
11156                                     NaN

### Делаем сравнение различных моделей

Описание всех моделей, заложенных в turicreate, доступно по ссылке:

https://apple.github.io/turicreate/docs/api/turicreate.toolkits.recommender.html

In [11]:
print('⬛⬛⬛⬛⬛⬛⬛---------------\nTrain default model')
model_0 = tc.recommender.create(training_data, 'user_id', 'group_id', target = 'log_visits',
                                user_data = users_1, item_data = groups_1)

print('⬛⬛⬛⬛⬛⬛⬛---------------\nTrain item_similarity model 1')
model_1 = tc.item_similarity_recommender.create(training_data, 'user_id', 'group_id', target = 'log_visits',
                                                user_data = users_1, item_data = groups_1)

print('⬛⬛⬛⬛⬛⬛⬛---------------\nTrain item_similarity model 2')
model_1 = tc.item_similarity_recommender.create(training_data, 'user_id', 'group_id',
                                                user_data = users_1, item_data = groups_1)


print('⬛⬛⬛⬛⬛⬛⬛---------------\nTrain popularity_recommender model')
model_2 = tc.popularity_recommender.create(training_data, 'user_id', 'group_id', target = 'log_visits',
                                           user_data = users_1, item_data = groups_1)

print('⬛⬛⬛⬛⬛⬛⬛---------------\nTrain factorization_recommender model')
model_3 = tc.factorization_recommender.create(training_data, 'user_id', 'group_id', target = 'log_visits',
                                              user_data = users_1, item_data = groups_1)

print('⬛⬛⬛⬛⬛⬛⬛---------------\nTrain ranking_factorization_recommender model')
model_4 = tc.ranking_factorization_recommender.create(training_data, 'user_id', 'group_id',
                                                      user_data = users_1, item_data = groups_1)

⬛⬛⬛⬛⬛⬛⬛---------------
Train default model


Preparing data set.

Data has 600249 observations with 50906 users and 24535 items.

Data prepared in: 0.74489s

Training ranking_factorization_recommender for recommendations.

+--------------------------------+--------------------------------------------------+----------+

| Parameter                      | Description                                      | Value    |

+--------------------------------+--------------------------------------------------+----------+

| num_factors                    | Factor Dimension                                 | 32       |

| regularization                 | L2 Regularization on Factors                     | 1e-09    |

| solver                         | Solver used for training                         | adagrad  |

| linear_regularization          | L2 Regularization on Linear Coefficients         | 1e-09    |

| ranking_regularization         | Rank-based Regularization Weight                 | 0.25     |

| side_data_factorization        | Assign Factors for Side Data                     | True     |

| max_iterations                 | Maximum Number of Iterations                     | 25       |

+--------------------------------+--------------------------------------------------+----------+

Optimizing model using SGD; tuning step size.

Using 75031 / 600249 points for tuning the step size.

+---------+-------------------+------------------------------------------+

| Attempt | Initial Step Size | Estimated Objective Value                |

+---------+-------------------+------------------------------------------+

| 0       | 2.63158           | Not Viable                               |

| 1       | 0.657895          | Not Viable                               |

| 2       | 0.164474          | Not Viable                               |

| 3       | 0.0411184         | 0.306334                                 |

| 4       | 0.0205592         | 0.421181                                 |

| 5       | 0.0102796         | 0.740924                                 |

+---------+-------------------+------------------------------------------+

| Final   | 0.0411184         | 0.306334                                 |

+---------+-------------------+------------------------------------------+

Starting Optimization.

+---------+--------------+-------------------+-----------------------+-------------+

| Iter.   | Elapsed Time | Approx. Objective | Approx. Training RMSE | Step Size   |

+---------+--------------+-------------------+-----------------------+-------------+

| Initial | 82us         | 2.77711           | 1.1903                |             |

+---------+--------------+-------------------+-----------------------+-------------+

| 1       | 680.229ms    | 1.35443           | 1.05173               | 0.0411184   |

| 2       | 1.24s        | 0.865796          | 0.831567              | 0.0411184   |

| 3       | 1.80s        | 0.759438          | 0.769795              | 0.0411184   |

| 4       | 2.34s        | 0.706076          | 0.737191              | 0.0411184   |

| 5       | 2.89s        | 0.673749          | 0.715786              | 0.0411184   |

| 10      | 5.49s        | 0.594523          | 0.661343              | 0.0411184   |

| 20      | 10.46s       | 0.527829          | 0.61232               | 0.0411184   |

| 25      | 12.95s       | 0.504069          | 0.595855              | 0.0411184   |

+---------+--------------+-------------------+-----------------------+-------------+

Optimization Complete: Maximum number of passes through the data reached.

Computing final objective value and training RMSE.

Final objective value: 0.480797

Final training RMSE: 0.573993

⬛⬛⬛⬛⬛⬛⬛---------------
Train item_similarity model 1


Warning: Ignoring columns направление 1, направление 2, weekend;

To use these columns in scoring predictions, use a model that allows the use of additional features.

Preparing data set.

Data has 600249 observations with 50906 users and 24535 items.

Data prepared in: 0.525575s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 806us                          | 1.75       |

| 3.399ms                        | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing data in one pass using dense lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 402.332ms                           | 0                | 0               |

| 863.487ms                           | 100              | 24535           |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

⬛⬛⬛⬛⬛⬛⬛---------------
Train item_similarity model 2

Finished training in 0.887295s

Warning: Ignoring columns log_visits, направление 1, направление 2, weekend;

To use one of these as a rating column, specify the column name to be used as target

and use a method that allows the use of a target.

Preparing data set.

Data has 600249 observations with 50906 users and 24535 items.

Data prepared in: 0.48523s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 672us                          | 1.75       |

| 3.976ms                        | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing data in one pass using dense lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 408.93ms                            | 0                | 0               |

| 923.399ms                           | 100              | 24535           |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

⬛⬛⬛⬛⬛⬛⬛---------------
Train popularity_recommender model

Finished training in 0.951201s

Warning: Ignoring columns направление 1, направление 2, weekend;

To use these columns in scoring predictions, use a model that allows the use of additional features.

Preparing data set.

Data has 600249 observations with 50906 users and 24535 items.

Data prepared in: 0.524392s

600249 observations to process; with 24535 unique items.

⬛⬛⬛⬛⬛⬛⬛---------------
Train factorization_recommender model


Preparing data set.

Data has 600249 observations with 50906 users and 24535 items.

Data prepared in: 0.667044s

Training factorization_recommender for recommendations.

+--------------------------------+--------------------------------------------------+----------+

| Parameter                      | Description                                      | Value    |

+--------------------------------+--------------------------------------------------+----------+

| num_factors                    | Factor Dimension                                 | 8        |

| regularization                 | L2 Regularization on Factors                     | 1e-08    |

| solver                         | Solver used for training                         | adagrad  |

| linear_regularization          | L2 Regularization on Linear Coefficients         | 1e-10    |

| side_data_factorization        | Assign Factors for Side Data                     | True     |

| max_iterations                 | Maximum Number of Iterations                     | 50       |

+--------------------------------+--------------------------------------------------+----------+

Optimizing model using SGD; tuning step size.

Using 75031 / 600249 points for tuning the step size.

+---------+-------------------+------------------------------------------+

| Attempt | Initial Step Size | Estimated Objective Value                |

+---------+-------------------+------------------------------------------+

| 0       | 2.63158           | Not Viable                               |

| 1       | 0.657895          | Not Viable                               |

| 2       | 0.164474          | Not Viable                               |

| 3       | 0.0411184         | 0.291593                                 |

| 4       | 0.0205592         | 0.555126                                 |

| 5       | 0.0102796         | 0.91922                                  |

+---------+-------------------+------------------------------------------+

| Final   | 0.0411184         | 0.291593                                 |

+---------+-------------------+------------------------------------------+

Starting Optimization.

+---------+--------------+-------------------+-----------------------+-------------+

| Iter.   | Elapsed Time | Approx. Objective | Approx. Training RMSE | Step Size   |

+---------+--------------+-------------------+-----------------------+-------------+

| Initial | 48us         | 1.4168            | 1.19029               |             |

+---------+--------------+-------------------+-----------------------+-------------+

| 1       | 187.573ms    | 0.978454          | 0.989147              | 0.0411184   |

| 2       | 377.223ms    | 0.759402          | 0.871409              | 0.0411184   |

| 3       | 556.532ms    | 0.695778          | 0.8341                | 0.0411184   |

| 4       | 733.447ms    | 0.660099          | 0.812429              | 0.0411184   |

| 5       | 911.507ms    | 0.634921          | 0.79678               | 0.0411184   |

| 10      | 1.77s        | 0.569631          | 0.75469               | 0.0411184   |

| 25      | 4.37s        | 0.50492           | 0.710512              | 0.0411184   |

| 50      | 8.68s        | 0.468012          | 0.684036              | 0.0411184   |

+---------+--------------+-------------------+-----------------------+-------------+

Optimization Complete: Maximum number of passes through the data reached.

Computing final objective value and training RMSE.

Final objective value: 0.456077

Final training RMSE: 0.675255

⬛⬛⬛⬛⬛⬛⬛---------------
Train ranking_factorization_recommender model


Preparing data set.

Data has 600249 observations with 50906 users and 24535 items.

Data prepared in: 0.732407s

Training ranking_factorization_recommender for recommendations.

+--------------------------------+--------------------------------------------------+----------+

| Parameter                      | Description                                      | Value    |

+--------------------------------+--------------------------------------------------+----------+

| num_factors                    | Factor Dimension                                 | 32       |

| regularization                 | L2 Regularization on Factors                     | 1e-09    |

| solver                         | Solver used for training                         | adagrad  |

| linear_regularization          | L2 Regularization on Linear Coefficients         | 1e-09    |

| binary_target                  | Assume Binary Targets                            | True     |

| side_data_factorization        | Assign Factors for Side Data                     | True     |

| max_iterations                 | Maximum Number of Iterations                     | 25       |

+--------------------------------+--------------------------------------------------+----------+

Optimizing model using SGD; tuning step size.

Using 75031 / 600249 points for tuning the step size.

+---------+-------------------+------------------------------------------+

| Attempt | Initial Step Size | Estimated Objective Value                |

+---------+-------------------+------------------------------------------+

| 0       | 2.5               | Not Viable                               |

| 1       | 0.625             | No Decrease (1.56868 >= 1.38667)         |

| 2       | 0.15625           | 0.281096                                 |

| 3       | 0.078125          | 0.298158                                 |

| 4       | 0.0390625         | 0.313403                                 |

| 5       | 0.0195312         | 0.350342                                 |

+---------+-------------------+------------------------------------------+

| Final   | 0.15625           | 0.281096                                 |

+---------+-------------------+------------------------------------------+

Starting Optimization.

+---------+--------------+-------------------+-----------------------------------+-------------+

| Iter.   | Elapsed Time | Approx. Objective | Approx. Training Predictive Error | Step Size   |

+---------+--------------+-------------------+-----------------------------------+-------------+

| Initial | 48us         | 1.38667           | 0.693789                          |             |

+---------+--------------+-------------------+-----------------------------------+-------------+

| 1       | 832.041ms    | 0.678453          | 0.240271                          | 0.15625     |

| 2       | 1.70s        | 0.404047          | 0.127614                          | 0.15625     |

| 3       | 2.58s        | 0.348939          | 0.107094                          | 0.15625     |

| 4       | 3.45s        | 0.322663          | 0.0977861                         | 0.15625     |

| 5       | 4.29s        | 0.303611          | 0.091835                          | 0.15625     |

| 10      | 8.69s        | 0.260453          | 0.0784377                         | 0.15625     |

| 15      | 12.91s       | 0.236882          | 0.0709593                         | 0.15625     |

| 20      | 17.20s       | 0.222573          | 0.0667345                         | 0.15625     |

| 25      | 21.44s       | 0.209652          | 0.0626152                         | 0.15625     |

+---------+--------------+-------------------+-----------------------------------+-------------+

Optimization Complete: Maximum number of passes through the data reached.

Computing final objective value and training Predictive Error.

Final objective value: 0.205477

Final training Predictive Error: 0.0607248

In [12]:
tc.recommender.util.compare_models(
    validation_data, 
    [model_0, model_1,model_2,model_3,model_4], 
    model_names=["⬛⬛⬛⬛⬛⬛⬛ Default model",
                 "⬛⬛⬛⬛⬛⬛⬛ Item similarity model",
                 '⬛⬛⬛⬛⬛⬛⬛ Popularity recommender',
                 '⬛⬛⬛⬛⬛⬛⬛ Factorization recommender',
                 '⬛⬛⬛⬛⬛⬛⬛ Ranking factorization recommender'])

PROGRESS: Evaluate model ⬛⬛⬛⬛⬛⬛⬛ Default model

Precision and recall summary statistics by cutoff
+--------+----------------+-------------+
| cutoff | mean_precision | mean_recall |
+--------+----------------+-------------+
|   1    |      0.0       |     0.0     |
|   2    |      0.0       |     0.0     |
|   3    |      0.0       |     0.0     |
|   4    |      0.0       |     0.0     |
|   5    |      0.0       |     0.0     |
|   6    |      0.0       |     0.0     |
|   7    |      0.0       |     0.0     |
|   8    |      0.0       |     0.0     |
|   9    |      0.0       |     0.0     |
|   10   |      0.0       |     0.0     |
+--------+----------------+-------------+
[10 rows x 3 columns]




Overall RMSE: 0.9807509470753443

Per User RMSE (best)
+-----------+---------------------+-------+
|  user_id  |         rmse        | count |
+-----------+---------------------+-------+
| 101410303 | 0.03251218011175805 |   1   |
+-----------+---------------------+-------+
[1 rows x 3 columns]


Per User RMSE (worst)
+-----------+-------------------+-------+
|  user_id  |        rmse       | count |
+-----------+-------------------+-------+
| 101386423 | 3.172759626878093 |   1   |
+-----------+-------------------+-------+
[1 rows x 3 columns]


Per Item RMSE (best)
+-----------+---------------------+-------+
|  group_id |         rmse        | count |
+-----------+---------------------+-------+
| 801366801 | 0.02801711718737221 |   1   |
+-----------+---------------------+-------+
[1 rows x 3 columns]


Per Item RMSE (worst)
+-----------+-------------------+-------+
|  group_id |        rmse       | count |
+-----------+-------------------+-------+
| 801348843 | 3.172759626878093 | 


Precision and recall summary statistics by cutoff
+--------+-----------------------+----------------------+
| cutoff |     mean_precision    |     mean_recall      |
+--------+-----------------------+----------------------+
|   1    |          0.0          |         0.0          |
|   2    |          0.0          |         0.0          |
|   3    |          0.0          |         0.0          |
|   4    |          0.0          |         0.0          |
|   5    |          0.0          |         0.0          |
|   6    |          0.0          |         0.0          |
|   7    | 0.0011904761904761904 | 0.008333333333333335 |
|   8    | 0.0010416666666666669 | 0.008333333333333335 |
|   9    |  0.001851851851851851 | 0.010416666666666668 |
|   10   | 0.0016666666666666668 | 0.010416666666666668 |
+--------+-----------------------+----------------------+
[10 rows x 3 columns]



[{'precision_recall_by_user': Columns:
  	user_id	int
  	cutoff	int
  	precision	float
  	recall	float
  	count	int
  
  Rows: 2160
  
  Data:
  +-----------+--------+-----------+--------+-------+
  |  user_id  | cutoff | precision | recall | count |
  +-----------+--------+-----------+--------+-------+
  | 101373685 |   1    |    0.0    |  0.0   |   1   |
  | 101373685 |   2    |    0.0    |  0.0   |   1   |
  | 101373685 |   3    |    0.0    |  0.0   |   1   |
  | 101373685 |   4    |    0.0    |  0.0   |   1   |
  | 101373685 |   5    |    0.0    |  0.0   |   1   |
  | 101373685 |   6    |    0.0    |  0.0   |   1   |
  | 101373685 |   7    |    0.0    |  0.0   |   1   |
  | 101373685 |   8    |    0.0    |  0.0   |   1   |
  | 101373685 |   9    |    0.0    |  0.0   |   1   |
  | 101373685 |   10   |    0.0    |  0.0   |   1   |
  +-----------+--------+-----------+--------+-------+
  [2160 rows x 5 columns]
  Note: Only the head of the SFrame is printed.
  You can use print_rows(nu

`Default model:             `Overall RMSE: 1.06 -> 1.01

`Item similarity model:     `Overall RMSE: 1.93 -> 1.98

`Popularity recommender:    `Overall RMSE: 0.98 -> 0.99

`Factorization recommender: `Overall RMSE: 1.00 -> 1.02

`Ranking factorization:     `Overall RMSE: 1.16 -> 0.97

### Сохраняем результаты Ranking factirozation модели

In [13]:
preds = model_4.recommend(k=10)
display(preds)

recommendations finished on 1000/50906 queries. users per second: 910.811

recommendations finished on 2000/50906 queries. users per second: 917.832

recommendations finished on 3000/50906 queries. users per second: 917.308

recommendations finished on 4000/50906 queries. users per second: 916.97

recommendations finished on 5000/50906 queries. users per second: 918.188

recommendations finished on 6000/50906 queries. users per second: 917.764

recommendations finished on 7000/50906 queries. users per second: 919.392

recommendations finished on 8000/50906 queries. users per second: 919.374

recommendations finished on 9000/50906 queries. users per second: 918.844

recommendations finished on 10000/50906 queries. users per second: 920.274

recommendations finished on 11000/50906 queries. users per second: 921.236

recommendations finished on 12000/50906 queries. users per second: 921.581

recommendations finished on 13000/50906 queries. users per second: 920.791

recommendations finished on 14000/50906 queries. users per second: 919.911

recommendations finished on 15000/50906 queries. users per second: 920.828

recommendations finished on 16000/50906 queries. users per second: 920.365

recommendations finished on 17000/50906 queries. users per second: 920.552

recommendations finished on 18000/50906 queries. users per second: 919.87

recommendations finished on 19000/50906 queries. users per second: 919.671

recommendations finished on 20000/50906 queries. users per second: 918.939

recommendations finished on 21000/50906 queries. users per second: 918.92

recommendations finished on 22000/50906 queries. users per second: 918.659

recommendations finished on 23000/50906 queries. users per second: 917.692

recommendations finished on 24000/50906 queries. users per second: 916.988

recommendations finished on 25000/50906 queries. users per second: 916.708

recommendations finished on 26000/50906 queries. users per second: 916.299

recommendations finished on 27000/50906 queries. users per second: 916.423

recommendations finished on 28000/50906 queries. users per second: 916.433

recommendations finished on 29000/50906 queries. users per second: 916.683

recommendations finished on 30000/50906 queries. users per second: 916.484

recommendations finished on 31000/50906 queries. users per second: 916.364

recommendations finished on 32000/50906 queries. users per second: 916.434

recommendations finished on 33000/50906 queries. users per second: 916.829

recommendations finished on 34000/50906 queries. users per second: 916.782

recommendations finished on 35000/50906 queries. users per second: 916.712

recommendations finished on 36000/50906 queries. users per second: 916.838

recommendations finished on 37000/50906 queries. users per second: 916.896

recommendations finished on 38000/50906 queries. users per second: 917.079

recommendations finished on 39000/50906 queries. users per second: 917.393

recommendations finished on 40000/50906 queries. users per second: 917.505

recommendations finished on 41000/50906 queries. users per second: 917.682

recommendations finished on 42000/50906 queries. users per second: 918.04

recommendations finished on 43000/50906 queries. users per second: 918.244

recommendations finished on 44000/50906 queries. users per second: 918.529

recommendations finished on 45000/50906 queries. users per second: 918.752

recommendations finished on 46000/50906 queries. users per second: 918.886

recommendations finished on 47000/50906 queries. users per second: 918.95

recommendations finished on 48000/50906 queries. users per second: 919.189

recommendations finished on 49000/50906 queries. users per second: 919.395

recommendations finished on 50000/50906 queries. users per second: 919.717

Columns:
	user_id	int
	group_id	int
	score	float
	rank	int

Rows: 509060

Data:
+-----------+-----------+--------------------+------+
|  user_id  |  group_id |       score        | rank |
+-----------+-----------+--------------------+------+
| 101354797 | 801353793 | 0.8232877948893121 |  1   |
| 101354797 | 801353285 | 0.8006826386368828 |  2   |
| 101354797 | 801355955 | 0.7825871354328703 |  3   |
| 101354797 | 801353290 | 0.7817690849324592 |  4   |
| 101354797 | 801353281 | 0.7737643818273108 |  5   |
| 101354797 | 801349127 | 0.7666385081690917 |  6   |
| 101354797 | 801349305 | 0.7456999964208675 |  7   |
| 101354797 | 801358792 | 0.7050070735630181 |  8   |
| 101354797 | 801348527 | 0.7040772802803721 |  9   |
| 101354797 | 801353708 | 0.6707442089823471 |  10  |
+-----------+-----------+--------------------+------+
[509060 rows x 4 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

### Сохраняем результаты для бэка

In [ ]:
preds.to_dataframe().to_csv('predicts/predict_v01.csv', index=False )

### Делаем предсказание для Test (с выборкой предсказаний на март-апрель 2023)

In [ ]:
import json
with open('groups2023.json') as json_file:
    groups2023 = json.load(json_file)

df_test = pd.read_csv('test.csv')
df_test.head()
df_test['уникальный номер участника'] = df_test['уникальный номер участника'].astype(int)


def display_history_and_predictions_for_user2(user_id, train_dataset, predictions):
    print('----------------------------------------------------------------------------')
    print('User', user_id, 'history:')
    #display(train_dataset[train_dataset['user_id'] == user_id]['group_id','is_online','session','направление 1','направление 2'].to_dataframe())
    display(train_dataset[train_dataset['user_id'] == user_id].to_dataframe())
    predictions = predictions[predictions['user_id'] == user_id]
    print('User recommendations:')
    predicted_groups = predictions['group_id'].to_list()
    display(df_groups[df_groups['уникальный номер'].isin(predicted_groups)].drop(columns=[
        'адрес площадки',
        'расписание в активных периодах',
        'расписание в закрытых периодах',
        'расписание в плановом периоде']))

In [ ]:
results = [] 


def get_groups_mix(user_id):
    # Predictions
    x0 = model_0.recommend(users=[user_id], k=1000).to_dataframe()
    x0 = x0[x0['group_id'].isin(groups2023)]
    x0 = x0.head(10)
    
    x1 = model_1.recommend(users=[user_id], k=1000).to_dataframe()
    x1 = x1[x1['group_id'].isin(groups2023)]
    x1 = x1.head(5)
    
    x2 = model_2.recommend(users=[user_id], k=1000).to_dataframe()
    x2 = x2[x2['group_id'].isin(groups2023)]
    x2 = x2.head(10)
    
    x3 = model_3.recommend(users=[user_id], k=1000).to_dataframe()
    x3 = x3[x3['group_id'].isin(groups2023)]
    x3 = x3.head(1)
    
    x4 = model_4.recommend(users=[user_id], k=1000).to_dataframe()
    x4 = x4[x4['group_id'].isin(groups2023)]
    x4 = x4.head(2)
    
    # Отбираем 10 штук, чтобы 1 была от 3 алгоритма и 2 от 4 алгоритма
    x = pd.concat([x1, x4, x0])
    x = x.drop_duplicates(subset='group_id').head(10) 
    
    return x
    
       

for index, row in df_test.iterrows():
    # Получаем юзера
    user_id = row['уникальный номер участника']
    
    x = get_groups_mix(user_id)
    
    display_history_and_predictions_for_user2(user_id, training_data, x)

    groups_list = x['group_id'].to_list()
    groups_list = [str(x) for x in groups_list]    
    groups_str = ','.join(groups_list)
    results.append(groups_str)
 

In [ ]:
df_test['уникальный номер группы'] = results
df_test.to_csv('test.csv', index=False)

In [ ]:
display(df_test)